### PR TITLE
[doc/otbn] Clear external CSRs

### PR DESCRIPTION
**Pull request by vrozic**
Specification for clearing INSN_CNT and ERR_BITS registers.

Signed-off-by: Vladimir Rozic <vrozic@lowrisc.org>

### CODE REVIEW

**Review by rswarbrick — COMMENTED** [review]
Is this in draft for any particular reason?

This looks good, but I'd suggest
1. Explicitly saying what happens if you try to write to one of the registers when OTBN is running (I think the write is ignored?)
2. Making the wording match for the two registers' clearing behaviour. Not a big deal, but we don't want a reader to think there might be some meaning behind the fact that they don't look the same.

File: hw/ip/otbn/data/otbn.hjson [hunk 1]
```diff
@@ -293,9 +293,14 @@
 
         Refer to the "List of Errors" section for a detailed description of the
         errors.
+
+        The host CPU can clear this register when OTBN is not running,
+        by writing any value.. Write attempts while OTBN is running are ignored.
```
Comment (rswarbrick) [inline hw/ip/otbn/data/otbn.hjson:298]: Nit: Double "." (and the same for insn_cnt below)

File: hw/ip/otbn/rtl/otbn.sv [hunk 2]
```diff
@@ -623,44 +627,44 @@ module otbn
   // ERR_BITS register
   // The error bits for an OTBN operation get stored on the cycle that done is
   // asserted. Software is expected to read them out before starting the next operation.
-  assign hw2reg.err_bits.bad_data_addr.de = done;
-  assign hw2reg.err_bits.bad_data_addr.d = err_bits.bad_data_addr;
 
-  assign hw2reg.err_bits.bad_insn_addr.de = done;
-  assign hw2reg.err_bits.bad_insn_addr.d = err_bits.bad_insn_addr;
+  assign hw2reg.err_bits.bad_data_addr.d = err_bits_q.bad_data_addr;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:631]: This is much cleaner than what was there before: nice! I don't think we need the blank lines between the assignments any more though (before, they were separating out blocks corresponding to the different registers). Would you mind dropping them?

File: hw/ip/otbn/rtl/otbn.sv [hunk 3]
```diff
@@ -623,44 +627,44 @@ module otbn
...
+  assign hw2reg.err_bits.illegal_bus_access.d = err_bits_q.illegal_bus_access;
 
-  assign hw2reg.err_bits.fatal_software.de = done;
-  assign hw2reg.err_bits.fatal_software.d = err_bits.fatal_software;
+  assign hw2reg.err_bits.lifecycle_escalation.d = err_bits_q.lifecycle_escalation;
+
+  assign hw2reg.err_bits.fatal_software.d = err_bits_q.fatal_software;
+
+  assign err_bits_clear = reg2hw.err_bits.bad_data_addr.qe;
+  assign err_bits_d = (err_bits_clear & otbn_inactive) ? '0 : err_bits;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:658]: Could you factor `otbn_inactive` into `err_bits_clear` here? I think it would be easier to understand (because otherwise it looks like we're setting a "clear me" flag and then not doing the clearing).

Similarly for `insn_cnt_clear` below, which I think would allow you to get rid of the "inactive" signal in the controller.

File: hw/ip/otbn/rtl/otbn.sv [hunk 4]
```diff
@@ -617,47 +621,44 @@ module otbn
...
-  assign hw2reg.err_bits.illegal_bus_access.d = err_bits.illegal_bus_access;
-
-  assign hw2reg.err_bits.lifecycle_escalation.de = done;
-  assign hw2reg.err_bits.lifecycle_escalation.d = err_bits.lifecycle_escalation;
-
-  assign hw2reg.err_bits.fatal_software.de = done;
-  assign hw2reg.err_bits.fatal_software.d = err_bits.fatal_software;
+  always_ff @(posedge clk_i or negedge rst_ni) begin
+    if (!rst_ni) begin
+      err_bits_q <= err_bits_zero;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:657]: Doesn't `'0` work here?

File: hw/ip/otbn/rtl/otbn.sv [hunk 5]
```diff
@@ -803,6 +806,12 @@ module otbn
     assign done = otbn_use_model ? done_rr_model : done_rtl;
     assign locked = otbn_use_model ? locked_model : locked_rtl;
     assign err_bits = otbn_use_model ? err_bits_model : err_bits_rtl;
+    assign err_bits_clear = otbn_use_model ? 1'b0 :
+                            (reg2hw.err_bits.bad_data_addr.qe & otbn_inactive);
+    assign err_bits_en = otbn_use_model ? done_rr_model :
+                         (err_bits_clear | err_bits_en_otbn);
+    assign err_bits_d = otbn_use_model ? err_bits_model :
+                        err_bits_clear ? err_bits_zero : err_bits_rtl;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:814]: This is a bit weird. The only remaining use of the original `err_bits` seems to be for setting `FATAL_ALERT_CAUSE`. I don't think this can be right: should it be dropped and replaced with `err_bits_q` maybe?

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 6]
```diff
@@ -521,7 +526,7 @@ module otbn_controller
   end
 
   always_comb begin
-    if (state_reset_i || state_q == OtbnStateLocked) begin
+    if (state_reset_i || state_q == OtbnStateLocked || (insn_cnt_clear_i && otbn_inactive_i)) begin
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn_controller.sv:529]: Not a big fan of this. Wouldn't it be easier to qualify the "clear" signal on "inactive" further up, rather than passing the two down together?

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 7]
```diff
@@ -474,6 +477,7 @@ module otbn_controller
         err_bits_q.fatal_software <= 1'b0;
         err_bits_q.lifecycle_escalation <= 1'b0;
         err_bits_q.illegal_bus_access  <= 1'b0;
+        err_bits_q.bad_internal_state <= 1'b0;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn_controller.sv:480]: Oh! That looks like I forgot to add it when I added the BAD_INTERNAL_STATE error. Could you split it into a separate commit? It can stay in the PR, but it's probably better to have unrelated bugfixes in different commits in the eventual history.

File: hw/ip/otbn/rtl/otbn.sv [hunk 8]
```diff
@@ -93,9 +93,13 @@ module otbn
   logic busy_execute_d, busy_execute_q;
   logic done;
   logic locked;
+  logic otbn_inactive;
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:96]: Can't you use `~ busy_execute_q` here? It avoids some unnecessary plumbing and I think also fixes an off-by-one timing problem:
image

File: hw/ip/otbn/rtl/otbn_controller.sv (hunk 6 above)
Comment (rswarbrick) [reply hw/ip/otbn/rtl/otbn_controller.sv:529]: Oh! Looks like I asked this before: https://github.com/lowRISC/opentitan/pull/9219#discussion_r756476657

File: hw/ip/otbn/rtl/otbn.sv (hunk 5 above)
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn.sv:814]: I think this is a bit backwards too, you want to first select `err_bits` choosing either model or rtl and then use that to feed your `err_bits_d`, i.e. something like

```systemverilog
assign err_bits_d = err_bits_clear ? '0 : err_bits
```

Then when using the model alone error bit clearing still works.

File: hw/ip/otbn/rtl/otbn_controller.sv (hunk 6 above)
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn_controller.sv:529]: You could introduce a new signal purely for clearing the instruction count which would factor in the external clear request along with `state_reset_i` and use that here instead.

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 9]
```diff
@@ -24,9 +24,11 @@ module otbn_controller
   input  logic  rst_ni,
 
   input  logic  start_i, // start the processing at address zero
+  input  logic  otbn_inactive_i, // OTBN is not running. Some CSRs can be cleared by the software.
   output logic  locked_o, // OTBN in locked state and must be reset to perform any further actions
 
   output err_bits_t err_bits_o, // valid when done_o is asserted
+  output logic      err_bits_en_o,
```
Comment (GregAC) [inline hw/ip/otbn/rtl/otbn_controller.sv:31]: I don't think we need this new output, we can just capture `err_bits_o` when we see `done_o`.

Comment (vrozic) [reply hw/ip/otbn/rtl/otbn_controller.sv]: OK, I have removed `err_bits_en_o` for now, but I might have to rework it later when enabling internal secure wipe, because `err_bits` would have to be registered before the `done_o` signal.

File: hw/ip/otbn/rtl/otbn.sv (hunk 4 above)
Comment (vrozic) [reply hw/ip/otbn/rtl/otbn.sv]: Indeed it does. I've changed it in the latest version.

File: hw/ip/otbn/rtl/otbn_controller.sv (hunk 7 above)
Comment (vrozic) [reply hw/ip/otbn/rtl/otbn_controller.sv]: This is no longer needed.
I've changed this to `err_bits_q = '0;`

File: hw/ip/otbn/rtl/otbn.sv (hunk 8 above)
Comment (vrozic) [reply hw/ip/otbn/rtl/otbn.sv]: Agreed and done

File: hw/ip/otbn/rtl/otbn.sv (hunk 5 above)
Comment (rswarbrick) [reply hw/ip/otbn/rtl/otbn.sv]: I think I'm convinced that the most recent version of this works when not using the model, but it's quite complicated! In particular, the path `err_bits_q -> hwreg.err_bits -> (... the RTL ...) -> err_bits_rtl -> err_bits -> err_bits_d` is not massively obvious! :-)

Could you add some comments next to the declarations of the signals, explaining what they all do?

I also suspect this won't work when using the model, which doesn't know to clear `err_bits` so there's no feedback from `err_bits_q -> hw2reg.err_bits -- !! nothing here !! -> err_bits_model -> err_bits -> err_bits_d`. Maybe `otbn_core_model` needs a `err_bits_clear_i` port?

File: hw/ip/otbn/rtl/otbn.sv [hunk 10]
```diff
@@ -786,6 +776,7 @@ module otbn
     assign done = otbn_use_model ? done_rr_model : done_rtl;
     assign locked = otbn_use_model ? locked_model : locked_rtl;
     assign err_bits = otbn_use_model ? err_bits_model : err_bits_rtl;
+    assign err_bits_en = otbn_use_model ? done_rr_model : (err_bits_clear | done_rtl);
```
Comment (rswarbrick) [inline hw/ip/otbn/rtl/otbn.sv:779]: I don't think that will work with the model enabled. Maybe this?
```
     assign err_bits_en = (otbn_use_model ? done_rr_model : done_rtl) | err_bits_clear;
```

File: hw/ip/otbn/rtl/otbn_controller.sv [hunk 11]
```diff
@@ -520,8 +511,10 @@ module otbn_controller
     end
   end
 
+  assign insn_cnt_clear = state_reset_i || state_q == OtbnStateLocked || insn_cnt_clear_i;
```
Comment (GregAC) [inline hw/ip/otbn/rtl/otbn_controller.sv:514]: nit: Should be single `|` as it's an assignment, not a conditional expression. Also I have a slight preference to place () around the `==` expression, might even be required for the `|` operator, cannot remember the precedence rules off hand (which is why I always put the brackets in ;) )

**Review by GregAC — APPROVED** [review]

**Comment by GregAC** [comment]
LGTM once you've addressed @rswarbrick's comment

File: hw/ip/otbn/rtl/otbn.sv (hunk 5 above)
Comment (vrozic) [reply hw/ip/otbn/rtl/otbn.sv]: I've changed this a bit. Now `err_bits_clear` is generated without using any signals from `otbn_core`.

`err_bits_clear = reg2hw.err_bits.bad_data_addr.qe & ~busy_execute_q;`
